lib/register.php

```diff
@@ -371,6 +371,42 @@ function gutenberg_content_has_blocks( $content ) {
 	return false !== strpos( $content, '<!-- wp:' );
 }
 
+/**
+ * Determine whether a post has a specific block. This test optimizes for
+ * performance rather than strict accuracy, detecting the pattern of a block
+ * but not validating its structure. For strict accuracy, you should use the
+ * block parser on post content.
+ *
+ * @see gutenberg_parse_blocks()
+ *
+ * @since 1.9
+ *
+ * @param object $post Post.
+ * @param string $block_name Full Block name to look for.
+ * @return bool Whether the content contain the specified block.
+ */
+function gutenberg_post_has_block( $post, $block_name ) {
+	$post = get_post( $post );
+	return $post && gutenberg_content_has_block( $post->post_content, $block_name );
+}
+
+/**
+ * Determine whether a content string contains a specific block. This test
+ * optimizes for performance rather than strict accuracy, detecting the
+ * pattern of a block but not validating its structure. For strict accuracy,
+ * you should use the block parser on post content.
+ *
+ * @since 1.9
+ * @see gutenberg_parse_blocks()
+ *
+ * @param string $content Content to test.
+ * @param string $block_name Full Block name to look for.
+ * @return bool Whether the content contain the specified block.
+ */
+function gutenberg_content_has_block( $content, $block_name ) {
+	return false !== strpos( $content, '<!-- wp:' . $block_name );
```

Comment:
There's a good chance for false positives here with substrings of block names, i.e.

```php
gutenberg_content_has_block( '<!-- wp:apples /-->', 'app' )
```

...is true.

(Aside: This would make a good test case to check against)

Reply:
Oh, very good point. Checking for whitespace after the block name would avoid this.

```php
return false !== strpos( $content, '<!-- wp:' . $block_name . ' ' );
```

Reply:
Shouldn't <code>gutenberg_content_has_block( '&lt;!--		&nbsp;&nbsp;&nbsp;&nbsp;wp:baba /-->', 'baba' )</code>,  also return `true`?

Reply:
Yeah. Regex seems to be warranted here if we're to get this done properly.

One step further:

```php
$pq_block_name = preg_quote( $block_name, '/' );
if ( preg_match( '/\<\!\-\-\s+wp\:' . $pq_block_name . '\s/', $content ) ) {
```

All the way:

```php
$doc = new DOMDocument();
$doc->loadHTML( $content );

$pq_block_name = preg_quote( $block_name, '/' );
$xpath         = new DOMXPath( $doc );

foreach ( $xpath->query( '//comment()' ) as $comment ) {
    $comment = $comment->textContent;

    if ( false !== strpos( $comment, 'wp:' . $block_name ) ) {
        if ( preg_match( '/^\s*wp\:' . $pq_block_name . '\s/', $comment ) ) {
            return true;
        }
    }
}
return false;
```

Reply:
Given the description of the function and the expected output of the block serializer, I could also argue that no it shouldn't.

>This test optimizes for performance rather than strict accuracy

In fairness, the parser _does_ support multiple spaces and perhaps this should too, but perhaps also the parser could be changed to be made more strict. I don't see any particularly compelling reason why it should need to support variations of these patterns.

Reply:
I think if the parser supports it, then this should too, because there's an existing function that like this one that [stands in front of the parser](https://github.com/WordPress/gutenberg/blob/8c5678c4735759263c7cc25ceb0d51cc30be3fd7/lib/blocks.php#L55-L60). However, changing the parser seems like an avenue worth exploring. I can't think of any reason that comment markers would lose their original whitespace formatting, but plugins do some interesting things sometimes.

Reply:
> In fairness, the parser does support multiple spaces and perhaps this should too, but perhaps also the parser could be changed to be made more strict. I don't see any particularly compelling reason why it should need to support variations of these patterns.

This goes against all my intuitions as a developer – I may want to indent things or put each attribute on a different line. Not to mention that outside tools have re-format the HTML and we wouldn't want valid HTML comments to be rejected by our parser. At our scale the chance of any of those happening actually isn't that low.

Reply:
I mean, how is this any different than me expecting the following to work? It won't.

```html
<  !    - 

  -    wp:foo   /  -  - >
```

Reply:
This isn't a valid HTML comment. If it were we should have supported it.

Reply:
>This isn't a valid HTML comment.

Right, this is my point. Just as other specifications can be strict on whitespace patterns, I see no reason why we should want or need to support superfluous spacing in our parser.

Reply:
Maybe looking at this with two different specifications in mind would help:

1. `<!-- -->` is the HTML comment specification.
2. `<!-- [ ] -->` Inside that is the Gutenberg block specification.

So if the parser ever changes, my opinion is that any changes there should not alter the way comment markers are parsed, as those are part of an established standard that Gutenberg has adopted, but should not alter the interpretation of.

Reply:
With regard to trailing whitespace, the parser expects [`/-->`](https://github.com/WordPress/gutenberg/blob/0dbab97c5af965fe32a62cc782980266b3b9891e/lib/parser.php#L1835) to be without whitespace. Continuing down a potential rabbit hole, maybe `<!-- wp:block / -->` should be allowed also.